### DISCUSSION TITLE
check if `query` or `mutation` is happening

### DISCUSSION THREAD

**Discussion by brenoliradev** [posva/pinia-colada#120] [Questions] [answered]
I want to retrieve if any query is fetching and if any mutation is mutating. Is there a way to do this?   

I'm currently using this: https://tanstack.com/query/latest/docs/reference/QueryClient#queryclientisfetching 

ps: if there isn't a way - and it would be a welcome feature - I could implement something using this as reference.

**Comment by posva** [comment]
This can be done by checking the query cache. I’m interested in knowing the use case as that would change how and if this needs to be exposed

**Comment by brenoliradev** [comment]
For me it's just a easy way to put a global check. 

I have a modal component that does multiple different mutations but each one of them has a different mutationKey since I'm using this pattern [automatic revalidation](https://tkdodo.eu/blog/automatic-query-invalidation-after-mutations) so It's easy to just check if there's _any_ mutation happening. 

I think the key role would be supporting the  [filters](https://tanstack.com/query/latest/docs/framework/react/guides/filters#query-filters). 

I think it's less relevant having an `isFetching` because `useQuery().isLoading` can be globally and _in tanstack_ `mutation.isMutating` isn't.

**Comment by posva** [comment, marked as answer]
FYI filters are supported. The `queryCache` is still not properly documented, so only the API is there. Relying on autocompletion helps but eventually the docs will be there.

For queries, you can do

```ts
const isLoading = computed(() => queryCache.getEntries({ predicate: (entry) => entry.asyncStatus.value === 'loading' }))
```

For mutations, this is not yet possible until the mutation store (#17) is exposed. Once exposed, they will have similar methods.

When it comes to invalidations, check [the docs](https://pinia-colada.esm.dev/guide/query-invalidation.html) guide on this. On the future it would be possible to pass a global `onSettled` option for mutations to invalidate everything (although I would highly advise against it).

If the purpose of these APIs are _global_, they are likely to be used one or two times in an app, so they will probably not be a dedicated API for it to promote using methods like `getEntries()` which are useful in many scenarios.

**Reply by brenoliradev** [reply]
Why do you highly advise against it? 

The revalidation pattern that I use only automatically revalidates the equivalent queryKey and I think it's a great pattern but would love to hear your opinion. 

About the  `getEntries` I strongly agree. It solves my problem and feels much more "vue" for me. I really like pinia-colada granularity.

**Reply by posva** [reply]
I wouldn't automatically invalidate everything on each mutation, but it's fine in small applications with only a few queries. Using `getEntries()` and then applying the different methods like `invalidate()`, `fetch()`, etc should give you a lot of flexibility already
